File: ios/Bilancio/Bilancio/Views/LaunchFlash.swift
//
//  LaunchFlash.swift
//  Bilancio
//
//  The owl, still, with the light moving across his chart.
//
//  It covers the moment the app is doing its least interesting work — fetching
//  the publishable key, configuring Clerk, restoring a session — so the wait
//  becomes the mark rather than a spinner. That it is also the brand arriving
//  is the point of it.
//
//  ── Why nothing moves but the light ────────────────────────────────────────
//
//  Chips used to fall onto the stacks, three to each. The idea was that the
//  brand assembled itself; what it looked like was a picture being built out of
//  parts, and every landing was one more chance to notice the join between a
//  chip and the drawing beneath it. A drawing that arrives whole has no joins
//  to notice.
//
//  So the drawing simply appears, and one band of lighter gold crosses the five
//  stacks from left to right. Each stack takes the light and hands it on, which
//  reads as the chart catching the light rather than as anything being built.
//  The band is masked to the stacks alone: sweeping the whole picture would
//  wash across the owl, and it is the money that should glint.
//
//  ── What `OwlLaunch` is ────────────────────────────────────────────────────
//
//  Not the app icon. The icon is a badge, and a badge shown large is a picture
//  of an icon rather than the owl himself. `OwlLaunch` is generated from the
//  same drawing with the plate taken off: cropped inside its bevel, its corners
//  rounded at the bevel's own radius so no part of the rim survives, and its
//  base faded out because the stacks run to the bottom of the drawing and a
//  hard edge there reads as the chart being cut off rather than standing on
//  the ground.
//
//  ── What `OwlField` is ─────────────────────────────────────────────────────
//
//  The gold around it, and the reason it cannot be seen to end. The drawing's
//  ground varies across it as well as down it, so a flat colour is wrong at the
//  sides and a vertical gradient is wrong in the middle — both leave the owl
//  sitting in a visible rectangle of slightly different gold.
//
//  `OwlField` is the drawing itself blurred until nothing but its ground is
//  left, then carried out to a canvas tall enough for any phone. A blurred copy
//  matches the ground at every point of the drawing's own edge because it is
//  made of it, and clamping it outwards leaves no streaks because there is
//  nothing sharp left in it to streak.
//

import SwiftUI

struct LaunchFlash: View {
    /// Set false to play the whole thing out and fade away.
    @Binding var showing: Bool

    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    @Environment(\.colorScheme) private var scheme

    /// The drawing fades in; the gold behind it never does. Fading the whole
    /// screen let the "Starting…" label underneath show through it.
    @State private var settled = false
    /// How lit each stack is, left to right.
    ///
    /// One plain rectangle per stack, and its opacity animated: three earlier
    /// attempts drew the light as a travelling band -- masked, then clipped
    /// inside a Canvas -- and none of them put a single pixel on screen. A
    /// rectangle at a position with an animated opacity is the least SwiftUI
    /// can be asked to do, and it is what the drawing's own fade already uses.
    @State private var glow = [Double](repeating: 0, count: Art.stacks.count)
    @State private var fading = false

    var body: some View {
        GeometryReader { geo in
            let art = Art(size: geo.size)

            ZStack {
                Image("OwlField")
                    .resizable()
                    .frame(width: art.fieldWidth, height: art.fieldHeight)
                    .position(x: geo.size.width / 2, y: geo.size.height / 2)

                Image("OwlLaunch")
                    .resizable()
                    .interpolation(.high)
                    .frame(width: art.width, height: art.height)
                    .position(x: geo.size.width / 2, y: geo.size.height / 2)
                    .opacity(settled ? 1 : 0)

                glint(in: art, across: geo.size)
            }
            .frame(width: geo.size.width, height: geo.size.height)
            // A full screen of gold is a lot of gold at night.
            .overlay(scheme == .dark ? Color.black.opacity(0.18) : .clear)
            .opacity(fading ? 0 : 1)
        }
        .ignoresSafeArea()
        .task { await play() }
        .accessibilityHidden(true)
    }

    /// The stacks, each taking the light and handing it on.
    @ViewBuilder
    private func glint(in art: Art, across screen: CGSize) -> some View {
        ForEach(Array(Art.stacks.indices), id: \.self) { index in
            let rect = art.stackRect(index)
            Rectangle()
                .fill(Self.sheen)
                .opacity(glow[index] * Self.peak)
                .frame(width: rect.width, height: rect.height)
                .position(x: rect.midX, y: rect.midY)
                .allowsHitTesting(false)
        }
    }

    /// How pale a stack goes at the top of its turn.
    private static let peak = 0.45

    /// Gold with the light on it, not white: white would grey the chart on its
    /// way past.
    private static let sheen = Color(red: 1.0, green: 0.94, blue: 0.74)

    private func play() async {
        // Reduce Motion is a request not to be moved at. The drawing still
        // arrives; the light simply does not travel across it.
        guard !reduceMotion else {
            settled = true
            try? await Task.sleep(for: .milliseconds(900))
            withAnimation(.easeOut(duration: 0.28)) { fading = true }
            try? await Task.sleep(for: .milliseconds(280))
            showing = false
            return
        }

        withAnimation(.easeOut(duration: 0.34)) { settled = true }

        /* Long enough for the drawing to actually be on screen.
         *
         * `task` runs when the view appears, which is before the launch screen
         * has finished handing over -- so a sweep started here played out while
         * the screen was still washing in from white, and was over by the time
         * there was anything to see it on. */
        try? await Task.sleep(for: .milliseconds(750))

        /* Left to right, each stack a breath behind the one before it. Up and
           down are scheduled together: the whole wave is set going in one pass
           and then simply waited out. */
        for index in Art.stacks.indices {
            let delay = Double(index) * 0.14
            withAnimation(.easeInOut(duration: 0.40).delay(delay)) { glow[index] = 1 }
            withAnimation(.easeInOut(duration: 0.45).delay(delay + 0.40)) { glow[index] = 0 }
        }
        try? await Task.sleep(for: .milliseconds(1_450))

        try? await Task.sleep(for: .milliseconds(260))
        withAnimation(.easeOut(duration: 0.3)) { fading = true }
        try? await Task.sleep(for: .milliseconds(300))
        showing = false
    }
}

// MARK: - Where the drawing sits

/// How the drawing meets the screen, and the gold behind it.
///
/// `OwlField` is generated from `OwlLaunch` at the same width, with the drawing
/// centred in it, so drawing both centred and at the same width puts every
/// point of the field exactly where it was blurred from.
private struct Art {
    /// `OwlLaunch`, in its own pixels.
    static let size = CGSize(width: 764, height: 824)
    /// `OwlField`, which is wider than any screen is tall relative to it.
    static let fieldSize = CGSize(width: 764, height: 1900)

    /// x range of each stack, left to right — its dark edges included, its
    /// drop shadow not. Measured against the outlines: the shadow beside each
    /// stack is close enough in tone to be taken for part of it, and taking it
    /// swells every stack and lights the ground beside it.
    static let stacks: [ClosedRange<CGFloat>] = [
        120...202, 219...315, 333...429, 450...549, 571...673,
    ]

    /// The painted top of each stack. The light starts there rather than at the
    /// top of the picture, so it crosses the chart and not the air above it.
    static let tops: [CGFloat] = [743, 681, 613, 562, 515]

    let size: CGSize

    /// The drawing spans the width. Anything larger crops the owl's wings and
    /// the outer stack; anything smaller leaves him small on a field of gold.
    var width: CGFloat { size.width }
    var height: CGFloat { width * Art.size.height / Art.size.width }

    /// The field, at the same scale as the drawing so the two line up.
    var fieldWidth: CGFloat { width }
    var fieldHeight: CGFloat { width * Art.fieldSize.height / Art.fieldSize.width }

    /// Points per artwork pixel.
    var scale: CGFloat { width / Art.size.width }

    /// Where the drawing starts on screen.
    var originY: CGFloat { (size.height - height) / 2 }

    /// One stack, in screen points.
    func stackRect(_ index: Int) -> CGRect {
        let stack = Art.stacks[index]
        let top = originY + Art.tops[index] * scale
        return CGRect(x: stack.lowerBound * scale,
                      y: top,
                      width: (stack.upperBound - stack.lowerBound) * scale,
                      height: originY + height - top)
    }
}

File: android/app/src/main/java/com/bilanciomoney/bilancio/ui/LaunchFlash.kt
package com.bilanciomoney.bilancio.ui

import android.provider.Settings
import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.FastOutSlowInEasing
import androidx.compose.animation.core.LinearOutSlowInEasing
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.ui.graphics.luminance
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.FilterQuality
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.imageResource
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.IntSize
import com.bilanciomoney.bilancio.R
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlin.math.roundToInt

/**
 * The owl, still, with the light moving across his chart -- the iPhone's
 * LaunchFlash, from the same two pictures and the same measurements.
 *
 * It covers the moment the app is doing its least interesting work -- Clerk
 * restoring the session -- so the wait becomes the mark rather than a spinner.
 * The drawing arrives whole and fades in; then one band of lighter gold crosses
 * the five stacks from left to right, each taking the light and handing it on,
 * so it is the money that glints and not the owl.
 *
 * `owl_launch` is the drawing with the icon's plate taken off; `owl_field` is
 * that drawing blurred until only its gold ground is left, carried out tall
 * enough for any phone, so the owl never sits in a visible rectangle. Both are
 * drawn at the full width of the screen and centred, which puts every point of
 * the field exactly where it was blurred from. See ios/.../LaunchFlash.swift.
 *
 * Drawn over the app rather than before it, so signing in carries on beneath.
 */
@Composable
fun LaunchFlash(onDone: () -> Unit) {
    val context = LocalContext.current
    val field = ImageBitmapRes(R.drawable.owl_field)
    val owl = ImageBitmapRes(R.drawable.owl_launch)
    /* The app's own appearance, which may differ from the phone's. */
    val dark = androidx.compose.material3.MaterialTheme.colorScheme.background.luminance() < 0.5f

    val settled = remember { Animatable(0f) }
    val fade = remember { Animatable(1f) }
    val glow = remember { List(STACKS.size) { Animatable(0f) } }

    LaunchedEffect(Unit) {
        /* Animations switched off on the phone are a request not to be moved
           at: the drawing still arrives, the light does not travel. */
        val still = Settings.Global.getFloat(context.contentResolver, Settings.Global.ANIMATOR_DURATION_SCALE, 1f) == 0f
        if (still) {
            settled.snapTo(1f)
            delay(900)
        } else {
            settled.animateTo(1f, tween(340, easing = LinearOutSlowInEasing))
            /* Long enough for the drawing to be seen before the light moves. */
            delay(750 - 340)
            /* Left to right, each stack a breath behind the one before. */
            glow.forEachIndexed { i, a ->
                launch {
                    delay(i * 140L)
                    a.animateTo(1f, tween(400, easing = FastOutSlowInEasing))
                    a.animateTo(0f, tween(450, easing = FastOutSlowInEasing))
                }
            }
            delay(1450 + 260)
        }
        fade.animateTo(0f, tween(300))
        onDone()
    }

    Canvas(Modifier.fillMaxSize()) {
        val w = size.width
        val scale = w / ART_W
        val artH = ART_H * scale
        val fieldH = FIELD_H * scale
        val originY = (size.height - artH) / 2
        val a = fade.value

        drawImage(
            field,
            dstOffset = IntOffset(0, ((size.height - fieldH) / 2).roundToInt()),
            dstSize = IntSize(w.roundToInt(), fieldH.roundToInt()),
            alpha = a,
            filterQuality = FilterQuality.High,
        )
        drawImage(
            owl,
            dstOffset = IntOffset(0, originY.roundToInt()),
            dstSize = IntSize(w.roundToInt(), artH.roundToInt()),
            alpha = settled.value * a,
            filterQuality = FilterQuality.High,
        )
        /* Gold with the light on it, not white, which would grey the chart. */
        STACKS.forEachIndexed { i, (from, to) ->
            val top = originY + TOPS[i] * scale
            drawRect(
                SHEEN.copy(alpha = glow[i].value * PEAK * a),
                Offset(from * scale, top),
                Size((to - from) * scale, originY + artH - top),
            )
        }
        /* A full screen of gold is a lot of gold at night. */
        if (dark) drawRect(Color.Black.copy(alpha = 0.18f * a))
    }
}

@Composable
private fun ImageBitmapRes(id: Int) = androidx.compose.ui.graphics.ImageBitmap.imageResource(id)

/* owl_launch and owl_field, in their own pixels. */
private const val ART_W = 764f
private const val ART_H = 824f
private const val FIELD_H = 1900f

/* x range of each stack, measured against the outlines (not the shadows), and
   the painted top of each -- so the light crosses the chart, not the air. */
private val STACKS = listOf(120f to 202f, 219f to 315f, 333f to 429f, 450f to 549f, 571f to 673f)
private val TOPS = listOf(743f, 681f, 613f, 562f, 515f)

private const val PEAK = 0.45f
private val SHEEN = Color(1.0f, 0.94f, 0.74f)
